odoo/fields.py

```diff
@@ -3070,6 +3075,676 @@ def _record_ids_per_res_model(self, records):
+
+        raise ValueError(f"Wrong property type {type(value)!r}")
+
+    # Record format: the value is a list, where each element is a dict
+    # containing the definition of a property, together with the property's
+    # corresponding value, like
+    #
+    #       [{
+    #           'name': '3adf37f3258cfe40f0907d9cbdd7d091',
+    #           'string': 'Color Code',
+    #           'type': 'char',
+    #           'default': 'blue',
+    #           'value': 'red',
+    #       }, {
+    #           'name': 'aa34746a6851ee4ea1f8d95746e45788',
+    #           'string': 'Partner',
+    #           'type': 'many2one',
+    #           'comodel': 'test_new_api.partner',
+    #           'value': 1337,
+    #       }]
+    #
+    def convert_to_record(self, value, record):
+        # value is in cache format
+        definition = self._get_properties_definition(record)
+        if not value or not definition:
+            return definition or []
+
+        assert isinstance(value, dict)
+        value = self._dict_to_list(value, definition)
+        self._parse_json_types(value, record.env)
+
+        return value
+
+    # Read format: almost identical to the record format, except that relational
+    # field values have a display name.
+    #
+    #       [{
+    #           'name': '3adf37f3258cfe40f0907d9cbdd7d091',
+    #           'string': 'Color Code',
+    #           'type': 'char',
+    #           'default': 'blue',
+    #           'value': 'red',
+    #       }, {
+    #           'name': 'aa34746a6851ee4ea1f8d95746e45788',
+    #           'string': 'Partner',
+    #           'type': 'many2one',
+    #           'comodel': 'test_new_api.partner',
+    #           'value': [1337, 'Bob'],
+    #       }]
+    #
+    def convert_to_read(self, value, record, use_name_get=True):
+        # value is in record format
+        if use_name_get:
+            self._add_display_name(value, record.env)
+        return value
+
+    def convert_to_write(self, value, record):
+        """If we write a list on the child, update the definition record."""
+        if isinstance(value, list):
+            # will update the definition record
+            self._remove_display_name(value)
+            return value
+
+        return super().convert_to_write(value, record)
+
+    def convert_to_onchange(self, value, record, names):
+        self._add_display_name(value, record.env)
+        return value
+
+    def write(self, records, value):
+        """Check if the properties definition has been changed.
+
+        To avoid extra SQL queries used to detect definition change, we add a
+        flag in the properties list. Parent update is done only when this flag
+        is present, delegating the check to the caller (generally web client).
+
+        For deletion, we need to keep the removed property definition in the
+        list to be able to put the delete flag in it. Otherwise we have no way
+        to know that a property has been removed.
```

Comment:
security-wise, any change of the definition deserve an explicit log message for traceability, as those schema changes will otherwise be nearly invisible, yet can entail data loss for users